libethereum/BlockChain.cpp

```diff
@@ -212,58 +211,70 @@ void BlockChain::init(ChainParams const& _p)
+        {
+            cnote << "Killing extras database (DB minor version:" << lastMinor << " != our miner version: " << c_minorProtocolVersion << ").";
+            DEV_IGNORE_EXCEPTIONS(fs::remove_all(extrasPath / fs::path("details.old")));
+            fs::rename(extrasPath / fs::path("extras"), extrasPath / fs::path("extras.old"));
+            fs::remove_all(extrasPath / fs::path("state"));
+            writeFile(extrasPath / fs::path("minor"), rlp(c_minorProtocolVersion));
+            lastMinor = (unsigned)RLP(status);
+        }
+
+        if (_we == WithExisting::Kill)
+        {
+            cnote << "Killing blockchain & extras database (WithExisting::Kill).";
+            fs::remove_all(chainPath / fs::path("blocks"));
+            fs::remove_all(extrasPath / fs::path("extras"));
+        }
     }
 
     try
     {
-        m_blocksDB.reset(new db::DBImpl(chainPath / fs::path("blocks")));
-        m_extrasDB.reset(new db::DBImpl(extrasPath / fs::path("extras")));
+        m_blocksDB = db::DBFactory::create(chainPath / fs::path("blocks"));
+        m_extrasDB = db::DBFactory::create(extrasPath / fs::path("extras"));
     }
     catch (db::DatabaseError const& ex)
     {
         // Check the exact reason of errror, in case of IOError we can display user-friendly message
         if (*boost::get_error_info<db::errinfo_dbStatusCode>(ex) != db::DatabaseStatus::IOError)
             throw;
 
-        if (fs::space(chainPath / fs::path("blocks")).available < 1024)
+        if (!db::isMemoryDB())
         {
-            cwarn << "Not enough available space found on hard drive. Please free some up and then re-run. Bailing.";
-            BOOST_THROW_EXCEPTION(NotEnoughAvailableSpace());
+            if (fs::space(chainPath / fs::path("blocks")).available < 1024)
+            {
+                cwarn << "Not enough available space found on hard drive. Please free some up and then re-run. Bailing.";
+                BOOST_THROW_EXCEPTION(NotEnoughAvailableSpace());
+            }
+            else
+            {
+                cwarn <<
+                    "Database " <<
+                    (chainPath / fs::path("blocks")) <<
+                    "or " <<
+                    (extrasPath / fs::path("extras")) <<
+                    "already open. You appear to have another instance of ethereum running. Bailing.";
+                BOOST_THROW_EXCEPTION(DatabaseAlreadyOpen());
+            }
         }
         else
         {
-            cwarn <<
-                "Database " <<
-                (chainPath / fs::path("blocks")) <<
-                "or " <<
-                (extrasPath / fs::path("extras")) <<
-                "already open. You appear to have another instance of ethereum running. Bailing.";
-            BOOST_THROW_EXCEPTION(DatabaseAlreadyOpen());
+            cwarn << "Unknown database error occurred during in-memory database creation";
+            BOOST_THROW_EXCEPTION(UnknownDatabaseError());
```

Comment:
I would `throw;` here, current exception `ex` could contain some valuable info from lower level, it's better to make it visible than to suppress it.